### PR TITLE
rgw: improve handling of illformed Swift's container ACLs.

### PR DESCRIPTION
**Pull request by rzarzynski**
Fixes: http://tracker.ceph.com/issues/18796
Signed-off-by: Radoslaw Zarzynski <rzarzynski@mirantis.com>

### CODE REVIEW

**Comment by rzarzynski** [comment]
Tempest hasn't found any regression here.

**Comment by rzarzynski** [comment]
@Jing-Scott: could you please take a look?

File: src/rgw/rgw_acl_swift.cc [hunk 1]
```diff
@@ -64,99 +66,123 @@ static bool uid_is_public(const string& uid)
...
+   * operation that is essential to preserve compliance with Swift. It con't
+   * be easily accomplished with boost::string_ref. */
   try {
+    bool is_negative;
+    ACLGrant grant;
+
     if ('-' == url_spec[0]) {
-      is_negative = true;
       url_spec = url_spec.substr(1);
+      boost::algorithm::trim_if(url_spec, boost::is_any_of(" \t\r\n"));
```
Comment (Jing-Scott) [inline src/rgw/rgw_acl_swift.cc]: the `boost::algorithm::trim_if` satisfied what I wanted before. very good!

File: src/rgw/rgw_acl_swift.cc [hunk 2]
```diff
@@ -64,99 +66,123 @@ static bool uid_is_public(const string& uid)
...
 
-      bool is_negative = false;
-      if (false == normalize_referer_urlspec(url_spec, is_negative)) {
-        ldout(cct, 10) << "cannot normalize referer: " << url_spec << dendl;
-        continue;
-      } else {
-        ldout(cct, 10) << "normalized referer to url_spec=" << url_spec
-                       << ", is_negative=" << is_negative << dendl;
-      }
+int RGWAccessControlPolicy_SWIFT::add_grants(RGWRados* const store,
```
Comment (Jing-Scott) [inline src/rgw/rgw_acl_swift.cc]: Yeah! This logic looks more clear and more easy to understand!

File: src/rgw/rgw_acl_swift.cc [hunk 3]
```diff
@@ -64,99 +66,123 @@ static bool uid_is_public(const string& uid)
          sub.compare(".referrer") == 0;
```
Comment (rzarzynski) [inline src/rgw/rgw_acl_swift.cc]: Absolutely.

**Comment by Jing-Scott** [comment]
@cbodley @mattbenjamin
I've no write access. So what do you think?

**Comment by rzarzynski** [comment]
@Jing-Scott: we definitely need to obtain results from Teuthology first. Just scheduled [a run](http://pulpito.ceph.com:80/rzarzynski-2017-02-04_13:09:50-rgw-wip-rzarzynski-testing---basic-smithi/) together with #13242.

**Comment by rzarzynski** [comment]
The Teuthology run is green except one but unrelated failure.

File: src/rgw/rgw_acl_swift.cc (hunk 1 above)
Comment (cbodley) [inline src/rgw/rgw_acl_swift.cc]: @rzarzynski if these strings are only passed as headers, i don't think we should be trying to handle control characters like `\t\r\n` here. looking at http://docs.openstack.org/developer/swift/overview_acl.html:
> Spaces may occur between elements as shown in the following example:
> 
> `.r : *, .rlistings, 7ec59e87c6584c348b563254aae4c221:*`

Comment (rzarzynski) [inline src/rgw/rgw_acl_swift.cc]: DONE.

**Comment by rzarzynski** [comment]
@cbodley: I've addressed your comment regarding the whitespace characters by switching from `boost::algorithm::trim_if` to plain `boost::algorithm::trim`. Absolutely no other changes here.

**Pull request merged**